python/lsst/psf/service/backend/_psf_service_backend.py

```diff
@@ -0,0 +1,338 @@
+# This file is part of psf_service_backend.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = ("PsfServiceBackend", "PsfExtraction")
+
+import dataclasses
+from uuid import UUID, uuid4
+from collections.abc import Sequence
+
+from lsst.afw.image import ImageD
+from lsst.daf.base import PropertyList
+from lsst.daf.butler import Butler, DataId, DatasetRef
+from lsst.resources import ResourcePath, ResourcePathExpression
+from .projection_finders import ProjectionFinder
+import lsst.geom as geom
+
+
+@dataclasses.dataclass
+class PsfExtraction:
+    """A struct that stores the extracted PSF model at a point."""
+
+    psf_image: ImageD
+    """The PSF image itself."""
+
+    metadata: PropertyList
+    """Additional FITS metadata about the PSF extraction process."""
+
+    origin_ref: DatasetRef
+    """Fully-resolved reference to the dataset the PSF is from."""
+
+    def write_fits(self, path: str) -> None:
+        """Write the PSF image to a FITS file.
+
+        Parameters
+        ----------
+        path : `str`
+            Local path to the file.
+        """
+        self.psf_image.writeFits(fileName=path, metadata=self.metadata)
+
+
+class PsfServiceBackend:
+    """High-level interface to the PSF service backend.
+
+    This backend can retrieve the PSF model from various LSST image datasets,
+    for example:
+    - `calexp` (Per-detector processed exposure),
+    - `deepCoadd_calexp` (calibrated coadd exposures),
+    - `deepDiff_differenceExp` (difference image produced in AP pipeline).
+
+    Any dataset type that has a `getPsf()` method should be supported. Refer to
+    the LSST Data Management code on GitHub (https://github.com/lsst) for more
+    information on dataset types.
+
+    Parameters
+    ----------
+    butler : `Butler`
```

Comment:
```suggestion
    butler : `lsst.daf.butler.Butler`
```
This needs to be fully-qualified to get linked properly in docs, if this ever gets a doc build.  Same is true of anything else you import from another package (I won't try to find all such cases here).

Reply:
Hmm I thought since I have already imported `Butler` using `from lsst.daf.butler import Butler`, I do not need to specify the full path (`lsst.daf.butler.Butler`) again in the docstring. But I guess it's more reliable if I specify the full path.

Reply:
Done! I also did it for all the other cases.

Reply:
Thanks.  As far as I can tell, Sphinx does not pay attention at all to how you have imported things (!)